### ISSUE TITLE
Arguments misordered for singularity-based kernels

### ISSUE THREAD

**Issue by zbarry** [NordicHPC/envkernel#4]
Thanks for creating this! It's been quite helpful. I'm using the 1.0.0 release.

It appears that the path to the Singularity image is added after `--bind` flags in the resulting `kernel.json` file. This results in an argument parsing error when Jupyter launches `envkernel`. Manually moving `"/path/to/image.sif"` to be right after `"singularity"` and `"run"` in the file resolves the problem.

The `envkernel` command:

```bash
envkernel singularity \
--name="$IMAGE_FULLNAME"_singularity --display-name="$IMAGE_FULLNAME (Singularity)" \
--user --contain --nv \
$SIF_PATH \
--bind /cm --bind /dlab --bind /scratch --bind /data --bind /home \
--bind /scratch/tmp:/tmp --bind /var --workdir /scratch/tmp
```

The `kernel.json` that is created:

```json
{
 "argv": [
  "/home/barryza1/miniconda3/envs/fibrosis/bin/envkernel",
  "singularity",
  "run",
  "--connection-file",
  "{connection_file}",
  "--contain",
  "--nv",
  "--bind",
  "/cm",
  "--bind",
  "/dlab",
  "--bind",
  "/scratch",
  "--bind",
  "/data",
  "--bind",
  "/home",
  "--bind",
  "/scratch/tmp:/tmp",
  "--bind",
  "/var",
  "--workdir",
  "/scratch/tmp",
  "/home/barryza1/computeenv/container_images/fibrosis-dev-0.7.sif",
  "--",
  "python",
  "-m",
  "ipykernel_launcher",
  "-f",
  "{connection_file}"
 ],
 "display_name": "fibrosis-dev-0.7 (Singularity)",
 "language": "python"
}
```

The error message:

```
[I 14:25:10.115 LabApp] Kernel started: ef486081-deac-41bc-9f99-d191625c0991
envkernel: cli args: ['--connection-file', '/home/barryza1/.local/share/jupyter/runtime/kernel-ef486081-deac-41bc-9f99-d191625c0991.json', '--contain', '--nv', '--bind', '/cm', '--bind', '/dlab', '--bind', '/scrat
ch', '--bind', '/data', '--bind', '/home', '--bind', '/scratch/tmp:/tmp', '--bind', '/var', '--workdir', '/scratch/tmp', '/home/barryza1/computeenv/container_images/fibrosis-dev-0.7.sif', '--', 'python', '-m', 'ip
ykernel_launcher', '-f', '/home/barryza1/.local/share/jupyter/runtime/kernel-ef486081-deac-41bc-9f99-d191625c0991.json']
run: args: Namespace(connection_file='/home/barryza1/.local/share/jupyter/runtime/kernel-ef486081-deac-41bc-9f99-d191625c0991.json', image='/cm', pwd=False)
run: unknown_args: ['--contain', '--nv', '--bind', '--bind', '/dlab', '--bind', '/scratch', '--bind', '/data', '--bind', '/home', '--bind', '/scratch/tmp:/tmp', '--bind', '/var', '--workdir', '/scratch/tmp', '/hom
e/barryza1/computeenv/container_images/fibrosis-dev-0.7.sif']
run: rest: ['python', '-m', 'ipykernel_launcher', '-f', '/home/barryza1/.local/share/jupyter/runtime/kernel-ef486081-deac-41bc-9f99-d191625c0991.json']
singularity: running cmd= singularity exec --bind /home/barryza1/.local/share/jupyter/runtime/kernel-ef486081-deac-41bc-9f99-d191625c0991.json:/kernel-ef486081-deac-41bc-9f99-d191625c0991.json --contain --nv --bin
d --bind /dlab --bind /scratch --bind /data --bind /home --bind /scratch/tmp:/tmp --bind /var --workdir /scratch/tmp /home/barryza1/computeenv/container_images/fibrosis-dev-0.7.sif /cm python -m ipykernel_launcher
 -f /kernel-ef486081-deac-41bc-9f99-d191625c0991.json
WARNING: skipping mount of /data/projects/notebooks/--bind: no such file or directory
WARNING: passwd file doesn't exist in container, not updating
WARNING: group file doesn't exist in container, not updating
FATAL:   container creation failed: mount /var/singularity/mnt/session/dev->/dev error: while mounting /var/singularity/mnt/session/dev: destination /dev doesn't exist in container
```

With this formulation of the JSON file:

```json
{
 "argv": [
  "/home/barryza1/miniconda3/envs/fibrosis/bin/envkernel",
  "singularity",
  "run",
  "--connection-file",
  "{connection_file}",
  "/home/barryza1/computeenv/container_images/fibrosis-dev-0.7.sif",
  "--contain",
  "--nv",
  "--bind",
  "/cm",
  "--bind",
  "/dlab",
  "--bind",
  "/scratch",
  "--bind",
  "/data",
  "--bind",
  "/home",
  "--bind",
  "/scratch/tmp:/tmp",
  "--bind",
  "/var",
  "--workdir",
  "/scratch/tmp",
  "--",
  "python",
  "-m",
  "ipykernel_launcher",
  "-f",
  "{connection_file}"
 ],
 "display_name": "fibrosis-dev-0.7 (Singularity)",
 "language": "python"
}
```

the kernel launches as expected.

**Comment by rkdarst** [comment]
I'm glad you are finding enkernel useful!

So, there are several things at play here, and I'm not sure exactly which problem is occuring here.

If you look at the logs, you notice two arguments `--bind --bind` right after each other - something got lost.  Because of the way Python's argparse handles unknown arguments, the first `--bind /cm` gets interperted as `--bind` as an argument, and the image filename as `/cm` (not the actual `.sif`).  I recommend using every argument as `--bind=/cm` (no spaces at all).

When you put the image filename first, then it detects the image right, but still puts the arguments as `--bind xxx` with the .sif file at the end, and same problem.  At one point I was investigating this but couldn't find a proper solution quickly enough, so I just left the solution at always all arguments with an `=` like `--bind=xxx`.

but... I'm still not entirely sure how this relates to the error message you saw.  Can you try adjusting arguments to this, and then see what happens?  Then, I'll see what comes next.  When I get back to it, I'll try to see if I can find an overall better solution to the problem.

**Comment by zbarry** [comment]
I can confirm that that fixes the problem. Thanks! The new, working `kernel.json` that envkernel creates just for reference is:

```json
{
 "argv": [
  "/home/barryza1/miniconda3/envs/fibrosis/bin/envkernel",
  "singularity",
  "run",
  "--connection-file",
  "{connection_file}",
  "--contain",
  "--nv",
  "--bind=/cm",
  "--bind=/dlab",
  "--bind=/scratch",
  "--bind=/data",
  "--bind=/home",
  "--bind=/scratch/tmp:/tmp",
  "--bind=/var",
  "--workdir=/scratch/tmp",
  "/home/barryza1/computeenv/container_images/fibrosis-dev-0.7.sif",
  "--",
  "python",
  "-m",
  "ipykernel_launcher",
  "-f",
  "{connection_file}"
 ],
 "display_name": "fibrosis-dev-0.7 (Singularity)",
 "language": "python"
}
```

**Comment by rkdarst** [comment]
I've also now created this branch as a test:
  https://github.com/NordicHPC/envkernel/tree/rkdarst/singularity-docker-image-order
which puts the image first in the kernel.json file, so it will be interpreted correctly no matter what the other arguments are.  But... I think it still would have gone wrong in the first phase (making the kernel.json) *if* the option with the space had been before the options without spaces...

I will leave this open until I can get around to testing better, then merge it.  Hopefully it removes at least a little bit of confusion for someone.  I also updated the docs to more explicitly warn of this problem.

Thanks for the feedback and improvements!
